### PR TITLE
class.newInstance is deprecated #2435

### PR DESCRIPTION
**Pull request by olamy**
Signed-off-by: olivier lamy <oliver.lamy@gmail.com>

### CODE REVIEW

**Comment by sbordet** [comment]
@olamy I'm not a big fan of multi catch.
This one is too much:

```java
catch (LoginException | IOException | UnsupportedCallbackException | InstantiationException
         | IllegalAccessException | NoSuchMethodException | InvocationTargetException | ClassNotFoundException e)
```

Most of the times, these multi catch block only surround the call to `newInstance()` and either we ignore (or log) the exception, or we rethrow a `RuntimeException`, so they can be simplified to:

```java
catch (Exception e)
```

Please get rid of multi catch blocks; if you really want to avoid wrapping `RuntimeException`s then:

```java
catch (RuntimeException x)
{
    throw x;
}
catch (Exception x)
{
    // log exception, or wrap it in RuntimeException, or keep existing code.
}
```

**Comment by olamy** [comment]
@sbordet definitely makes sense I made few changes.

File: jetty-ant/src/main/java/org/eclipse/jetty/ant/JettyRunTask.java [hunk 1]
```diff
@@ -147,13 +148,9 @@ public void setRequestLog(String className)
...
         {
-            this.requestLog = (RequestLog) Class.forName(className).newInstance();
+            this.requestLog = (RequestLog) Class.forName(className).getDeclaredConstructor().newInstance();
         }
-        catch (InstantiationException e)
-        {
-            throw new BuildException("Request logger instantiation exception: " + e);
-        }
-        catch (IllegalAccessException e)
+        catch (InstantiationException | NoSuchMethodException | InvocationTargetException | IllegalAccessException e)
```
Comment (sbordet) [inline jetty-ant/src/main/java/org/eclipse/jetty/ant/JettyRunTask.java]: Convert this block to `catch (Exception e)` and put it after the `catch (ClassNotFoundException e)` below.

File: jetty-jaas/src/main/java/org/eclipse/jetty/jaas/JAASLoginService.java [hunk 2]
```diff
@@ -19,6 +19,7 @@
 package org.eclipse.jetty.jaas;
 
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-jaas/src/main/java/org/eclipse/jetty/jaas/JAASLoginService.java]: Unused import ?

File: jetty-jaas/src/main/java/org/eclipse/jetty/jaas/spi/JDBCLoginModule.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.jaas.spi;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-jaas/src/main/java/org/eclipse/jetty/jaas/spi/JDBCLoginModule.java]: Unused import ?

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/utils/internal/PackageAdminServiceTracker.java [hunk 4]
```diff
@@ -280,24 +280,12 @@ private void invokeFragmentActivators(ServiceReference sr)
...
-            catch (InstantiationException e)
-            {
-                // e.printStackTrace();
-            }
-            catch (IllegalAccessException e)
-            {
-                // e.printStackTrace();
-            }
-            catch (ClassNotFoundException e)
+            catch (NullPointerException | InstantiationException | IllegalAccessException | ClassNotFoundException e)
```
Comment (sbordet) [inline jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/utils/internal/PackageAdminServiceTracker.java]: Remove this catch block and leave the `catch (Exception e)` below.

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java [hunk 5]
```diff
@@ -1284,14 +1285,15 @@ public void service(ServletRequest req, ServletResponse res) throws ServletExcep
      * @throws IllegalAccessException if not allowed to create a new instance
      * @throws InstantiationException if creating new instance resulted in error
      */
-    protected Servlet newInstance() throws ServletException, IllegalAccessException, InstantiationException
+    protected Servlet newInstance() throws ServletException, IllegalAccessException, InstantiationException,
+        NoSuchMethodException, InvocationTargetException
```
Comment (sbordet) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java]: Adding more exception to the signature is not enough.
You should A) update the javadocs too; B) update the catch block below, where `ServletException` is unwrapped.

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java [hunk 6]
```diff
@@ -98,7 +99,7 @@ public IncludeExcludeSet()
                 _excludePredicate = new SetContainsPredicate(_excludes);
             }
         }
-        catch (InstantiationException | IllegalAccessException e)
+        catch (Exception e)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: Add `catch (RuntimeException e)` before `catch (Exception e)` and rethrow without wrapping it.

File: jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/ClientContainer.java [hunk 7]
```diff
@@ -19,6 +19,7 @@
 package org.eclipse.jetty.websocket.jsr356;
 
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/ClientContainer.java]: Unused import ?

File: jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/DecoderFactory.java [hunk 8]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.jsr356;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/DecoderFactory.java]: Unused import ?

File: jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/EncoderFactory.java [hunk 9]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.jsr356;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/EncoderFactory.java]: Unused import ?

File: jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/client/AnnotatedClientEndpointConfig.java [hunk 10]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.jsr356.client;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/client/AnnotatedClientEndpointConfig.java]: Unused import ?

File: jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/AnnotatedServerEndpointConfig.java [hunk 11]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.jsr356.server;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/AnnotatedServerEndpointConfig.java]: Unused import ?

File: jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/ContainerDefaultConfigurator.java [hunk 12]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.jsr356.server;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/ContainerDefaultConfigurator.java]: Unused import ?

File: jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/deploy/WebSocketServerContainerInitializer.java [hunk 13]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.jsr356.server.deploy;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/deploy/WebSocketServerContainerInitializer.java]: Unused import ?

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/WebSocketExtensionFactory.java [hunk 14]
```diff
@@ -25,6 +25,8 @@
 import org.eclipse.jetty.websocket.api.extensions.ExtensionFactory;
 import org.eclipse.jetty.websocket.common.scopes.WebSocketContainerScope;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/WebSocketExtensionFactory.java]: Unused import ?

File: jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/NativeWebSocketConfiguration.java [hunk 15]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.websocket.server;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/NativeWebSocketConfiguration.java]: Unused import ?

File: jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java [hunk 16]
```diff
@@ -19,6 +19,7 @@
 package org.eclipse.jetty.websocket.server;
 
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: Unused import ?

**Comment by olamy** [comment]
@sbordet If I use my IDE for auto import this will reorder all imports and I'm pretty sure you might not be happy with that :P

File: jetty-jaas/src/main/java/org/eclipse/jetty/jaas/JAASLoginService.java [hunk 17]
```diff
@@ -228,7 +228,7 @@ else if (callback instanceof RequestParameterCallback)
             else
             {
                 Class<?> clazz = Loader.loadClass(_callbackHandlerClass);
-                callbackHandler = (CallbackHandler)clazz.newInstance();
+                callbackHandler = (CallbackHandler)clazz.getDeclaredConstructor().newInstance();
```
Comment (janbartel) [inline jetty-jaas/src/main/java/org/eclipse/jetty/jaas/JAASLoginService.java]: If I could get someone to review PR #2381 then this wouldn't be necessary :)

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/utils/internal/PackageAdminServiceTracker.java [hunk 18]
```diff
@@ -280,27 +280,11 @@ private void invokeFragmentActivators(ServiceReference sr)
...
-            {
-                // e.printStackTrace();
-            }
-            catch (ClassNotFoundException e)
-            {
-                // e.printStackTrace();
-            }
             catch (Exception e)
             {
                 e.printStackTrace();
```
Comment (janbartel) [inline jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/utils/internal/PackageAdminServiceTracker.java]: Ooops, I'll fix that printStackTrace and the other one in this file.

File: jetty-ant/src/main/java/org/eclipse/jetty/ant/JettyRunTask.java [hunk 19]
```diff
@@ -19,6 +19,7 @@
 package org.eclipse.jetty.ant;
 
 import java.io.File;
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-ant/src/main/java/org/eclipse/jetty/ant/JettyRunTask.java]: Unused import ?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 20]
```diff
@@ -21,6 +21,7 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java]: Unused import ?

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java [hunk 21]
```diff
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.util;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: Unused import ?

**Comment by sbordet** [comment]
@olamy your IDE reorders imports only if you configured it with the wrong order. Or get a better IDE :)

**Pull request merged**